File: android/expoview/src/main/java/host/exp/exponent/notifications/model/ScopedNotificationRequest.java
package host.exp.exponent.notifications.model;

import android.os.Parcel;

import androidx.annotation.Nullable;
import expo.modules.notifications.notifications.interfaces.NotificationTrigger;
import expo.modules.notifications.notifications.model.NotificationContent;
import expo.modules.notifications.notifications.model.NotificationRequest;
import host.exp.exponent.kernel.ExperienceKey;

public class ScopedNotificationRequest extends NotificationRequest {
  // We store String instead of ExperienceKey because ScopedNotificationRequest must be serializable.
  private String mExperienceScopeKeyString;

  public ScopedNotificationRequest(String identifier, NotificationContent content, NotificationTrigger trigger, @Nullable String experienceScopeKey) {
    super(identifier, content, trigger);
    mExperienceScopeKeyString = experienceScopeKey;
  }

  private ScopedNotificationRequest(Parcel in) {
    super(in);
    mExperienceScopeKeyString = in.readString();
  }

  public boolean checkIfBelongsToExperience(@Nullable ExperienceKey experienceKey) {
    if (mExperienceScopeKeyString == null) {
      return true;
    }
    return mExperienceScopeKeyString.equals(experienceKey.getScopeKey());
  }

  public static final Creator<ScopedNotificationRequest> CREATOR = new Creator<ScopedNotificationRequest>() {
    public ScopedNotificationRequest createFromParcel(Parcel in) {
      return new ScopedNotificationRequest(in);
    }

    public ScopedNotificationRequest[] newArray(int size) {
      return new ScopedNotificationRequest[size];
    }
  };

  @Nullable
  public String getExperienceScopeKeyString() {
    return mExperienceScopeKeyString;
  }

  @Override
  public void writeToParcel(Parcel dest, int flags) {
    super.writeToParcel(dest, flags);
    dest.writeString(mExperienceScopeKeyString);
  }
}

File: apps/expo-go/android/expoview/src/main/java/host/exp/exponent/notifications/model/ScopedNotificationRequest.kt
package host.exp.exponent.notifications.model

import expo.modules.notifications.notifications.model.NotificationRequest
import expo.modules.notifications.notifications.interfaces.NotificationTrigger
import android.os.Parcel
import host.exp.exponent.kernel.ExperienceKey
import android.os.Parcelable
import expo.modules.notifications.notifications.interfaces.INotificationContent

class ScopedNotificationRequest : NotificationRequest {
  // We store String instead of ExperienceKey because ScopedNotificationRequest must be serializable.
  val experienceScopeKeyString: String?

  constructor(
    identifier: String,
    content: INotificationContent,
    trigger: NotificationTrigger?,
    experienceScopeKey: String?
  ) : super(identifier, content, trigger) {
    experienceScopeKeyString = experienceScopeKey
  }

  private constructor(parcel: Parcel) : super(parcel) {
    experienceScopeKeyString = parcel.readString()
  }

  fun checkIfBelongsToExperience(experienceKey: ExperienceKey): Boolean {
    return if (experienceScopeKeyString == null) {
      true
    } else {
      experienceScopeKeyString == experienceKey.scopeKey
    }
  }

  override fun writeToParcel(dest: Parcel, flags: Int) {
    super.writeToParcel(dest, flags)
    dest.writeString(experienceScopeKeyString)
  }

  companion object CREATOR : Parcelable.Creator<ScopedNotificationRequest> {
    override fun createFromParcel(parcel: Parcel): ScopedNotificationRequest {
      return ScopedNotificationRequest(parcel)
    }

    override fun newArray(size: Int): Array<ScopedNotificationRequest?> {
      return arrayOfNulls(size)
    }
  }
}
